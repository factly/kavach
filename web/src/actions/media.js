import axios from 'axios';
import {
  ADD_MEDIUM,
  ADD_MEDIA,
  ADD_MEDIA_REQUEST,
  SET_MEDIA_LOADING,
  RESET_MEDIA,
  MEDIA_API,
} from '../constants/media';
import { buildObjectOfItems } from '../utils/objects';

import { addErrorNotification, addSuccessNotification } from './notifications';

export const getMedia = (query) => {
  return (dispatch) => {
    dispatch(loadingMedia());
    return axios
      .get(MEDIA_API, {
        params: query,
      })
      .then((response) => {
        dispatch(addMediaList(response.data.nodes));
        dispatch(
          addMediaRequest({
            data: response.data.nodes.map((item) => item.id),
            query: query,
            total: response.data.total,
          }),
        );
        dispatch(stopMediaLoading());
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const getMedium = (id) => {
  return (dispatch) => {
    dispatch(loadingMedia());
    return axios
      .get(MEDIA_API + '/' + id)
      .then((response) => {
        dispatch(getMediumByID(response.data));
        dispatch(stopMediaLoading());
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const addMedium = (data) => {
  return (dispatch) => {
    dispatch(loadingMedia());
    return axios
      .post(MEDIA_API, data)
      .then(() => {
        dispatch(resetMedia());
        dispatch(addSuccessNotification('Media Added'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const updateMedium = (data) => {
  return (dispatch) => {
    dispatch(loadingMedia());
    return axios
      .put(MEDIA_API + '/' + data.id, data)
      .then((response) => {
        dispatch(getMediumByID(response.data));
        dispatch(stopMediaLoading());
        dispatch(addSuccessNotification('Media Updated'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const deleteMedium = (id) => {
  return (dispatch) => {
    dispatch(loadingMedia());
    return axios
      .delete(MEDIA_API + '/' + id)
      .then(() => {
        dispatch(resetMedia());
        dispatch(addSuccessNotification('Media Deleted'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const addMedia = (media) => {
  return (dispatch) => {
    dispatch(addMediaList(media));
  };
};

export const loadingMedia = () => ({
  type: SET_MEDIA_LOADING,
  payload: true,
});

export const stopMediaLoading = () => ({
  type: SET_MEDIA_LOADING,
  payload: false,
});

export const getMediumByID = (data) => ({
  type: ADD_MEDIUM,
  payload: data,
});

export const addMediaList = (data) => (dispatch) => {
  dispatch({
    type: ADD_MEDIA,
    payload: buildObjectOfItems(data),
  });
};

export const addMediaRequest = (data) => ({
  type: ADD_MEDIA_REQUEST,
  payload: data,
});

export const resetMedia = () => ({
  type: RESET_MEDIA,
});
