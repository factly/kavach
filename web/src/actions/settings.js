import { TOGGLE_SIDER } from '../constants/settings';

export const toggleSider = () => {
  return (dispatch, getState) => {
    dispatch({
      type: TOGGLE_SIDER,
    });
  };
};
