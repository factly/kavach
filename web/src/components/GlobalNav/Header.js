import React from 'react';
import { useSelector } from 'react-redux';
import { Layout, Button, Popover, List, Avatar } from 'antd';
import { PlusOutlined } from '@ant-design/icons';
import SettingsIcon from '../Settings/img.svg';
import AccountMenu from './AccountMenu';
import OrganisationSelector from './OrganisationSelector';
import { Link } from 'react-router-dom';

function Header() {
  const { apps, organisationCount } = useSelector((state) => {
    const appIDs = state.organisations.details[state.organisations.selected]?.applications || [];
    return {
      apps:
        state.organisations.selected > 0
          ? appIDs.map((id) => ({
              ...state.applications.details[id],
              medium: state.media.details?.[state.applications.details[id]?.medium_id],
            }))
          : [],
      organisationCount: state.organisations.ids ? state.organisations.ids.length : 0,
    };
  });
  return (
    <Layout.Header className="layout-header">
      <div style={{ display: 'flex', justifyContent: 'flex-end', alignItems: 'center' }}>
        <div style={{ display: 'flex', height: '32px', gap: '14px' }}>
          {window.REACT_APP_ENABLE_MULTITENANCY === 'true' ? (
            <>
              <Link style={{ display: 'flex' }} to="/organisation/create">
                <Button icon={<PlusOutlined />} style={{ background: '#F1F1F1', border: 'none' }}>
                  New Organisation
                </Button>
              </Link>
              <OrganisationSelector />
            </>
          ) : organisationCount < 1 ? (
            <Link to="/organisation/create">
              <Button icon={<PlusOutlined />} style={{ background: '#F1F1F1', border: 'none' }}>
                New Organisation
              </Button>
            </Link>
          ) : null}
          {apps.length > 0 ? (
            <>
              <Popover
                placement="bottom"
                content={
                  <List
                    grid={{
                      gutter: 16,
                      xs: 1,
                      sm: 2,
                      md: 4,
                      lg: 4,
                      xl: 6,
                      xxl: 3,
                    }}
                    dataSource={apps}
                    renderItem={(item) => (
                      <List.Item>
                        <a
                          href={item.url}
                          style={{ textDecoration: 'none', color: 'inherit' }}
                          target="_blank"
                          rel="noopener noreferrer"
                        >
                          {item.medium && item.medium.url ? (
                            <img
                              alt="logo"
                              className="menu-logo"
                              src={
                                window.REACT_APP_ENABLE_IMGPROXY
                                  ? item.medium.url.proxy
                                  : item.medium.url?.raw
                              }
                            />
                          ) : (
                            <Avatar shape="square" size={35}>
                              {item.name?.charAt(0)}
                            </Avatar>
                          )}
                          <p>{item.name}</p>
                        </a>
                      </List.Item>
                    )}
                  />
                }
                trigger="click"
              >
                <img src={SettingsIcon} alt="icon" />
              </Popover>
            </>
          ) : null}
        </div>
        <AccountMenu style={{ float: 'right' }} />
      </div>
    </Layout.Header>
  );
}

export default Header;
