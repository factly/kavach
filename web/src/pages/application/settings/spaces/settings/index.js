import React from 'react';
import { useParams } from 'react-router-dom';
import SettingsList from '../../../../../components/Settings';

export default function SpaceSettings() {
  const { appID, spaceID } = useParams();
  return <SettingsList type="space" appID={appID} spaceID={spaceID} />;
}
