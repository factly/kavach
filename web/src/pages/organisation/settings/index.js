import React from 'react';
import { useSelector } from 'react-redux';
import SettingsList from '../../../components/Settings';
import { Skeleton } from 'antd';
export default function OrganisationSettings({ orgID }) {
  const { role, loading } = useSelector((state) => {
    return {
      role: state.profile.roles[state.organisations.selected],
      loading: state.profile.loading,
    };
  });
  return (
    <div>
      {
      (loading) ? <Skeleton/> :<SettingsList type="organisation" orgID={orgID} role={role}/>
      }
    </div>
  );
}
