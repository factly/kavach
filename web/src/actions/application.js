import axios from 'axios';
import {
  ADD_APPLICATION,
  ADD_APPLICATIONS,
  ADD_APPLICATIONS_REQUEST,
  SET_APPLICATIONS_LOADING,
  RESET_APPLICATIONS,
  APPLICATIONS_API,
  ADD_SPACE_IDS,
  SET_DEFAULT_APPLICATION_LOADING,
  GET_DEFAULT_APPLICATIONS,
} from '../constants/application';
import { ADD_APPLICATION_IDS, ORGANISATIONS_API } from '../constants/organisations';
import { buildObjectOfItems, deleteKeys, getIds, getValues } from '../utils/objects';
import { addMedia, addMediaList } from './media';
import { addErrorNotification, addSuccessNotification } from './notifications';
import { addSpaces } from './space';
import { addApplicationTokens } from './token';
import { addUsersList } from './users';
export const getApplications = () => {
  return (dispatch, getState) => {
    dispatch(loadingApplications());
    return axios
      .get(APPLICATIONS_API + '/' + getState().organisations.selected + '/applications')
      .then((response) => {
        dispatch(addApplicationList(response.data));
        dispatch(addApplicationIds(getIds(response.data)));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      })
      .finally(() => {
        dispatch(stopApplicationLoading());
      });
  };
};

export const addDefaultApplication = (appID) => {
  return (dispatch, getState) => {
    dispatch(loadingApplications());
    return axios
      .post(
        ORGANISATIONS_API +
          '/' +
          getState().organisations.selected +
          '/applications/' +
          appID +
          '/default',
      )
      .then(() => {
        dispatch(addSuccessNotification('Application Added Successfully'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      })
      .finally(() => {
        dispatch(stopApplicationLoading());
      });
  };
};

export const removeDefaultApplication = (appID) => {
  return (dispatch, getState) => {
    dispatch(loadingApplications());
    return axios
      .delete(
        ORGANISATIONS_API +
          '/' +
          getState().organisations.selected +
          '/applications/' +
          appID +
          '/default',
      )
      .then(() => {
        dispatch(addSuccessNotification('Application removed succesfully'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      })
      .finally(() => {
        dispatch(stopApplicationLoading());
      });
  };
};

export const getDefaultApplications = () => {
  return (dispatch, getState) => {
    dispatch(setDefaultApplicationsLoading());
    return axios
      .get(ORGANISATIONS_API + '/' + getState().organisations.selected + '/applications/default')
      .then((response) => {
        dispatch(addDefaultApplicationsList(response.data));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      })
      .finally(() => {
        dispatch(stopDefaultApplicationLoading());
      });
  };
};

export const getApplication = (id) => {
  return (dispatch, getState) => {
    dispatch(loadingApplications());
    return axios
      .get(APPLICATIONS_API + '/' + getState().organisations.selected + '/applications/' + id)
      .then((response) => {
        if (response.data.medium_id) {
          dispatch(addMedia(response.data.medium));
        }
        deleteKeys([response.data], ['medium']);
        dispatch(addUsersList(response.data.users));
        response.data.users = getIds(response.data.users);
        response.data.spaces = getIds(response.data.spaces)
        response.data.tokens = getIds(response.data.tokens)
        const spaces = getValues([response.data], 'spaces');
        // dispatch(addSpaces(spaces));
        dispatch(addApplication(response.data));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      })
      .finally(() => {
        dispatch(stopApplicationLoading());
      });
  };
};

export const createApplication = (data) => {
  data.is_default = false;
  return (dispatch, getState) => {
    dispatch(loadingApplications());
    return axios
      .post(ORGANISATIONS_API + '/' + getState().organisations.selected + '/applications', data)
      .then(() => {
        dispatch(resetApplications());
        dispatch(addSuccessNotification('Application Added'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const updateApplication = (data) => {
  return (dispatch, getState) => {
    dispatch(loadingApplications());
    return axios
      .put(
        ORGANISATIONS_API + '/' + getState().organisations.selected + '/applications/' + data.id,
        data,
      )
      .then(() => {
        dispatch(addSuccessNotification('Application Updated'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      })
      .finally(() => {
        dispatch(stopApplicationLoading());
      });
  };
};

export const deleteApplication = (id) => {
  return (dispatch, getState) => {
    dispatch(loadingApplications());
    return axios
      .delete(ORGANISATIONS_API + '/' + getState().organisations.selected + '/applications/' + id)
      .then(() => {
        dispatch(resetApplications());
        dispatch(addSuccessNotification('Application Deleted'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const addApplications = (applications) => {
  return (dispatch) => {
    dispatch(
      addMediaList(
        applications
          .filter((application) => application.medium)
          .map((application) => application.medium),
      ),
    );
    dispatch(
      addApplicationList(
        applications.map((application) => {
          return { ...application, medium: application.medium?.id };
        }),
      ),
    );
  };
};

export const loadingApplications = () => ({
  type: SET_APPLICATIONS_LOADING,
  payload: true,
});

export const stopApplicationLoading = () => ({
  type: SET_APPLICATIONS_LOADING,
  payload: false,
});

export const getApplicationByID = (data) => ({
  type: ADD_APPLICATION,
  payload: data,
});

export const addApplicationList = (data) => (dispatch) => {
  dispatch(loadingApplications());
  const medium = getValues(data, 'medium');
  dispatch(addMediaList(medium));
  deleteKeys(data, ['medium']);
  const spaces = getValues(data, 'spaces');
  dispatch(addSpaces(spaces));

  data.forEach((application) => {
    application.spaces = getIds(application.spaces);
    application.users = getIds(application.users);
    dispatch(addApplicationTokens(application.id, application.tokens));
    application.tokens = getIds(application.tokens);
  });

  dispatch({
    type: ADD_APPLICATIONS,
    payload: buildObjectOfItems(data),
  });
  dispatch(stopApplicationLoading());
};

export const addApplicationsRequest = (data) => ({
  type: ADD_APPLICATIONS_REQUEST,
  payload: data,
});

export const resetApplications = () => ({
  type: RESET_APPLICATIONS,
});

export const addApplicationIds = (data) => ({
  type: ADD_APPLICATION_IDS,
  payload: data,
});

export const addApplication = (data) => ({
  type: ADD_APPLICATION,
  payload: data,
});

export const addSpaceIDs = (appID, data) => ({
  type: ADD_SPACE_IDS,
  payload: {
    appID: appID,
    data: data,
  },
});

export const setDefaultApplicationsLoading = () => ({
  type: SET_DEFAULT_APPLICATION_LOADING,
  payload: true,
});

export const stopDefaultApplicationLoading = () => ({
  type: SET_DEFAULT_APPLICATION_LOADING,
  payload: false,
});

export const addDefaultApplicationsList = (data) => ({
  type: GET_DEFAULT_APPLICATIONS,
  payload: data,
});
