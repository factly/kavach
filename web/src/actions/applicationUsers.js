import axios from 'axios';
import {
  ADD_APPLICATION_USER,
  ADD_APPLICATION_USERS,
  ADD_APPLICATION_USERS_REQUEST,
  SET_APPLICATION_USERS_LOADING,
  RESET_APPLICATION_USERS,
  APPLICATION_USERS_API,
  ADD_USER_IDS,
} from '../constants/applicationUser';
import { getIds } from '../utils/objects';
import { addMediaList } from './media';
import { addErrorNotification, addSuccessNotification } from './notifications';
import { loadingUsers, stopUsersLoading } from './users';

export const getApplications = (appID) => {
  return (dispatch, getState) => {
    dispatch(loadingApplications());
    return axios
      .get(
        APPLICATION_USERS_API +
          '/' +
          getState().organisations.selected +
          '/applications/' +
          appID +
          '/users',
      )
      .then((response) => {
        dispatch(
          addApplicationList(
            response.data.users.map((user) => {
              return user;
            }),
          ),
        );
        dispatch(
          addApplicationsRequest({
            data: response.data.users.map((item) => item.id),
            application_id: response.data.application.id,
            total: response.data.total,
          }),
        );
        dispatch(stopApplicationLoading());
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const getApplication = (id) => {
  return (dispatch, getState) => {
    dispatch(loadingApplications());
    return axios
      .get(
        APPLICATION_USERS_API +
          '/' +
          getState().organisations.selected +
          '/applications/' +
          id +
          '/users',
      )
      .then((response) => {
        if (response.data.medium) dispatch(addMediaList([response.data.medium]));
        dispatch(getApplicationByID({ ...response.data, medium: response.data.medium?.id }));
        dispatch(stopApplicationLoading());
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const addApplicationUser = (data) => {
  return (dispatch, getState) => {
    dispatch(loadingApplications());
    return axios
      .post(
        APPLICATION_USERS_API +
          '/' +
          getState().organisations.selected +
          '/applications/' +
          data.application_id +
          '/users',
        data,
      )
      .then(() => {
        dispatch(resetApplications());
        dispatch(addSuccessNotification('Application User Added'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const updateApplication = (data) => {
  return (dispatch, getState) => {
    dispatch(loadingApplications());
    return axios
      .put(
        APPLICATION_USERS_API +
          '/' +
          getState().organisations.selected +
          '/applications/' +
          data.id,
        data,
      )
      .then((response) => {
        if (response.data.medium) dispatch(addMediaList([response.data.medium]));
        dispatch(getApplicationByID({ ...response.data, medium: response.data.medium?.id }));
        dispatch(stopApplicationLoading());
        dispatch(addSuccessNotification('Application Updated'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const deleteApplication = (id, appID) => {
  return (dispatch, getState) => {
    dispatch(loadingApplications());
    return axios
      .delete(
        APPLICATION_USERS_API +
          '/' +
          getState().organisations.selected +
          '/applications/' +
          id +
          '/users/' +
          appID,
      )
      .then(() => {
        dispatch(resetApplications());
        dispatch(addSuccessNotification('Application User Deleted'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const getApplicationUsers = (appID) => {
  return (dispatch, getState) => {
    dispatch(loadingUsers());
    return axios
      .get(
        APPLICATION_USERS_API +
          '/' +
          getState().organisations.selected +
          '/applications/' +
          appID +
          '/users',
      )
      .then((res) => {
        dispatch(addUserIds(getIds(res.data.users), appID));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      })
      .finally(() => {
        dispatch(stopUsersLoading());
      });
  };
};

export const addApplications = (applications) => {
  return (dispatch) => {
    dispatch(
      addMediaList(
        applications
          .filter((application) => application.medium)
          .map((application) => application.medium),
      ),
    );
    dispatch(
      addApplicationList(
        applications.map((application) => {
          return { ...application, medium: application.medium?.id };
        }),
      ),
    );
  };
};

export const loadingApplications = () => ({
  type: SET_APPLICATION_USERS_LOADING,
  payload: true,
});

export const stopApplicationLoading = () => ({
  type: SET_APPLICATION_USERS_LOADING,
  payload: false,
});

export const getApplicationByID = (data) => ({
  type: ADD_APPLICATION_USER,
  payload: data,
});

export const addApplicationList = (data) => ({
  type: ADD_APPLICATION_USERS,
  payload: data,
});

export const addApplicationsRequest = (data) => ({
  type: ADD_APPLICATION_USERS_REQUEST,
  payload: data,
});

const addUserIds = (data, appID) => ({
  type: ADD_USER_IDS,
  payload: {
    id: appID,
    data: data,
  },
});
export const resetApplications = () => ({
  type: RESET_APPLICATION_USERS,
});
