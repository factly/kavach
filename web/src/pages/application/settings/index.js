import React from 'react';
import { Link, useParams } from 'react-router-dom';
import SettingsList, { SettingsCard } from '../../../components/Settings';
import { Col, Row, Skeleton, Descriptions, Divider, Space } from 'antd';
import { ApartmentOutlined } from '@ant-design/icons';
import { getApplication } from '../../../actions/application';
import { useDispatch, useSelector } from 'react-redux';

export default function ApplicationSettings() {
  const { id } = useParams();
  const dispatch = useDispatch();
  const descriptionSpan = 3;
  const { application, loadingApp, role, loading } = useSelector((state) => {
    return {
      application: state.applications.details[id],
      loadingApp: state.applications.loading,
      role: state.profile.roles[state.organisations.selected],
      loading: state.profile.loading,
    };
  });

  React.useEffect(() => {
    dispatch(getApplication(id));
    //eslint-disable-next-line
  }, [id]);

  return (
    <div>
      {loading || loadingApp ? (
        <Skeleton />
      ) : (
        <Space direction="vertical" style={{ width: '100%' }}>
          <Descriptions title={<h2> Manage Application </h2>} bordered={true}>
            <Descriptions.Item label="Name" span={descriptionSpan}>
              {application?.name}
            </Descriptions.Item>
            <Descriptions.Item label="Description" span={descriptionSpan}>
              {application?.description}
            </Descriptions.Item>
            <Descriptions.Item label="Number of Spaces" span={descriptionSpan}>
              {application?.spaces?.length}
            </Descriptions.Item>
            <Descriptions.Item label="Number of Tokens" span={descriptionSpan}>
              {application?.tokens?.length}
            </Descriptions.Item>
            <Descriptions.Item label="Number of Tokens" span={descriptionSpan}>
              <a href={`https://${application?.url}`}>{application?.url}</a>
            </Descriptions.Item>
          </Descriptions>
          <Divider> Application Settings</Divider>
          <SettingsList type="application" appID={id} role={role} />
          <Row style={{ marginTop: 16 }}>
            <Col span={12}>
              <Link to={`/applications/${id}/settings/spaces`}>
                <SettingsCard
                  icon={<ApartmentOutlined style={{ color: '#4E89FF' }} />}
                  title="Spaces"
                  description="Space Settings"
                />
              </Link>
            </Col>
          </Row>
        </Space>
      )}
    </div>
  );
}
