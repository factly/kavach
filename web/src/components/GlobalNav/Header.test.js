import React from 'react';
import { useHistory } from 'react-router-dom';
import { BrowserRouter as Router } from 'react-router-dom';
import { useDispatch, Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mount, shallow } from 'enzyme';

import '../../matchMedia.mock';
import Header from './Header';
import { MenuUnfoldOutlined, MenuFoldOutlined } from '@ant-design/icons';
import * as actions from '../../actions/settings';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: jest.fn(),
}));

jest.mock('../../actions/settings', () => ({
  toggleSider: jest.fn(),
}));

describe('Header component', () => {
  let store;
  let mockedDispatch;

  let state;

  beforeEach(() => {
    state = {
      settings: {
        navTheme: 'dark',
        title: 'Dega',
        sider: {
          collapsed: true,
        },
      },
      organisations: {
        ids: [1],
        details: {
          1: {
            id: 1,
            title: 'title',
            description: 'description',
            permission: { role: 'member' },
            applications: [1, 2],
          },
        },
        loading: false,
        selected: 1,
      },
      applications: {
        details: {
          1: {
            id: 1,
            name: 'app1',
            description: 'description',
            permission: { role: 'member' },
            medium_id: 2,
          },
          2: {
            id: 2,
            name: 'app2',
            description: 'description',
            permission: { role: 'member' },
            medium_id: 1,
          },
        },
      },
      profile: {
        details: {
          id: '1',
          first_name: 'abc',
          last_name: 'xyz',
          display_name: 'abc',
          slug: 'abc',
          email: 'abc@gmail.com',
          social_media_urls: {},
          description: 'Description',
        },
        loading: false,
      },
      media: {
        details: {
          1: { id: 1, url: { proxy: 'imageUrl' } },
          2: { id: 2, url: { proxy: 'imageUrl' } },
        },
      },
    };
    store = mockStore(() => state);
    store.dispatch = jest.fn(() => ({}));
  });
  mockedDispatch = jest.fn(() => Promise.resolve({}));
  useDispatch.mockReturnValue(mockedDispatch);

  describe('snapshot testing', () => {
    it('should render the component', () => {
      const tree = shallow(
        <Provider store={store}>
          <Router>
            <Header />
          </Router>
        </Provider>,
      );
      expect(tree).toMatchSnapshot();
    });
  });
  describe('component testing', () => {
    let wrapper;
    it('should render correctly when organisations is present and when REACT_APP_ENABLE_MULTITENANCY is false', () => {
      window.REACT_APP_ENABLE_MULTITENANCY = 'false';
      wrapper = mount(
        <Provider store={store}>
          <Router>
            <Header />
          </Router>
        </Provider>,
      );

      expect(wrapper.find('Link').length).toBe(0);
    });
    it('should render correctly when organisations is present and when REACT_APP_ENABLE_MULTITENANCY is true', () => {
      window.REACT_APP_ENABLE_MULTITENANCY = 'true';
      wrapper = mount(
        <Provider store={store}>
          <Router>
            <Header />
          </Router>
        </Provider>,
      );
      expect(wrapper.find('Link').length).toBe(1);
    });
    it('should render correctly when organisations is not present and when REACT_APP_ENABLE_MULTITENANCY is false', () => {
      window.REACT_APP_ENABLE_MULTITENANCY = 'false';
      store = mockStore(() => {
        return {
          ...state,
          organisations: {
            details: {},
            loading: false,
            selected: 1,
          },
        };
      });

      wrapper = mount(
        <Provider store={store}>
          <Router>
            <Header />
          </Router>
        </Provider>,
      );

      expect(wrapper.find('Link').length).toBe(1);
    });
    it('should show application menu if applications not null and REACT_APP_ENABLE_IMGPROXY is enabled', () => {
      state = {
        settings: {
          navTheme: 'dark',
          title: 'Dega',
          sider: {
            collapsed: true,
          },
        },
        organisations: {
          ids: [1],
          details: {
            1: {
              id: 1,
              title: 'title',
              description: 'description',
              permission: { role: 'member' },
              applications: [1, 2],
            },
          },
          loading: false,
          selected: 1,
        },
        applications: {
          details: {
            1: {
              id: 1,
              name: 'app1',
              description: 'description',
              permission: { role: 'member' },
              medium_id: 2,
            },
            2: {
              id: 2,
              name: 'app2',
              description: 'description',
              permission: { role: 'member' },
              medium_id: 1,
            },
          },
        },
        profile: {
          details: {
            id: '1',
            first_name: 'abc',
            last_name: 'xyz',
            display_name: 'abc',
            slug: 'abc',
            email: 'abc@gmail.com',
            social_media_urls: {},
            description: 'Description',
          },
          loading: false,
        },
        media: {
          details: {
            1: { id: 1, url: { proxy: 'imageUrl', raw: 'rawurl' } },
            2: { id: 2 },
          },
        },
      };
      window.REACT_APP_ENABLE_IMGPROXY = 'true';
      store = mockStore(() => state);
      const wrapper = mount(
        <Provider store={store}>
          <Router>
            <Header />
          </Router>
        </Provider>,
      );
      wrapper.find('Popover').simulate('click');
      wrapper.update();
      expect(wrapper.find('img')).toHaveLength(1);
      expect(wrapper.find('img').prop('src')).toBe('imageUrl');
      expect(wrapper.find('Avatar').length).toBe(1);
      expect(wrapper.find('Avatar').prop('shape')).toBe('square');
      expect(wrapper.find('Avatar').text()).toBe('a');
      expect(wrapper.find('List').find('a').length).toBe(2);
    });
    it('should show application menu if applications not null and REACT_APP_ENABLE_IMGPROXY is disable and one of app do not have url', () => {
      state = {
        settings: {
          navTheme: 'dark',
          title: 'Dega',
          sider: {
            collapsed: true,
          },
        },
        organisations: {
          ids: [1],
          details: {
            1: {
              id: 1,
              title: 'title',
              description: 'description',
              permission: { role: 'member' },
              applications: [1, 2],
            },
          },
          loading: false,
          selected: 1,
        },
        applications: {
          details: {
            1: {
              id: 1,
              name: 'app1',
              description: 'description',
              permission: { role: 'member' },
              medium_id: 2,
            },
            2: {
              id: 2,
              name: 'app2',
              description: 'description',
              permission: { role: 'member' },
              medium_id: 1,
            },
          },
        },
        profile: {
          details: {
            id: '1',
            first_name: 'abc',
            last_name: 'xyz',
            display_name: 'abc',
            slug: 'abc',
            email: 'abc@gmail.com',
            social_media_urls: {},
            description: 'Description',
          },
          loading: false,
        },
        media: {
          details: {
            1: { id: 1, url: { proxy: 'imageUrl', raw: 'rawurl' } },
            2: { id: 2 },
          },
        },
      };
      window.REACT_APP_ENABLE_IMGPROXY = false;
      store = mockStore(() => state);
      const wrapper = mount(
        <Provider store={store}>
          <Router>
            <Header />
          </Router>
        </Provider>,
      );

      wrapper.find('Popover').simulate('click');
      wrapper.update();
      expect(wrapper.find('img')).toHaveLength(1);
      expect(wrapper.find('img').prop('src')).toBe('rawurl');
      expect(wrapper.find('List').find('a').length).toBe(2);
    });
    it('should not show application menu if no organisation selected', () => {
      state = {
        settings: {
          navTheme: 'dark',
          title: 'Dega',
          sider: {
            collapsed: true,
          },
        },
        organisations: {
          ids: [1],
          details: {
            1: {
              id: 1,
              title: 'title',
              description: 'description',
              permission: { role: 'member' },
              applications: [
                { id: 1, name: 'Dega', url: 'http://1233434/1323' },
                { id: 2, name: 'Bindu', url: 'http://1233434/454567' },
              ],
            },
          },
          loading: false,
          selected: 0,
        },
        profile: {
          details: {
            id: '1',
            first_name: 'abc',
            last_name: 'xyz',
            display_name: 'abc',
            slug: 'abc',
            email: 'abc@gmail.com',
            social_media_urls: {},
            description: 'Description',
          },
          loading: false,
        },
      };
      store = mockStore(() => state);
      const wrapper = mount(
        <Provider store={store}>
          <Router>
            <Header />
          </Router>
        </Provider>,
      );
      expect(wrapper.find('Popover').length).toBe(0);
    });
  });
});
