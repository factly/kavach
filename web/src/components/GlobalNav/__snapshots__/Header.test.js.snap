// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component snapshot testing should render the component 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [MockFunction],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <BrowserRouter>
    <Header />
  </BrowserRouter>
</ContextProvider>
`;
